learn-data-structures-by-building-a-shopping-cart/index.js

```diff
@@ -1,97 +1,207 @@
+    price: 11.99,
+    category: "macaroon",
+  },
+  {
+    id: 12,
+    name: "Lemon cupcakes (4 pack)",
+    price: 12.99,
+    category: "cupcake",
+  },
 ];
 
+// another good instance to show how map and destructuring works
+products.map(
+  ({ name, id, price }) =>
+    (dessertCards.innerHTML += `
+    <div class="dessert-card">
+      <h2>${name}</h2>
+      <p class="dessert-price">$${price}</p>
+      <button 
+        id="${id}" 
+        class="btn add-to-cart-btn">Add to cart
+      </button>
+    </div>
+  `)
+);
+
+// This might be the first time where classes are introduced
 class ShoppingCart {
+  // this might be the first time the this keyword is introduced
   constructor() {
     this.items = [];
     this.taxRate = 8.25;
-    this.discountPercentage = 30;
   }
 
   addItem(id, products) {
-    const item = products.find(item => item.id === id);
-    this.items.push(item);
-  }
+    // this might be the first time where the find method is introduced
+    const product = products.find((item) => item.id === id);
+    const { name, price } = product;
+    this.items.push(product);
+
+    const totalCountPerProduct = {};
+    this.items.forEach((dessert) => {
+      totalCountPerProduct[dessert.name] =
+        (totalCountPerProduct[dessert.name] || 0) + 1;
+    });
+
+    let currentProductCount = totalCountPerProduct[product.name];
+    let currentProductCountSpan = document.getElementById(
+      "product-count-for-id" + id
+    );
 
-  getItems() {
-    return this.items;
+    currentProductCount > 1
+      ? (currentProductCountSpan.textContent = `${currentProductCount}x`)
+      : (productsContainer.innerHTML += `
+      <div id=dessert${id} class="product">
+        <p>
+          <span class="product-count" id=product-count-for-id${id}></span>${name}
+        </p>
+        <p>$${price}</p>
+      </div>
+    `);
   }
 
   getCount() {
     return this.items.length;
   }
 
-  removeItem(id) {
-    const index = this.items.indexOf(id);
-    this.items.splice(index, 1);
-  }
-
   clearCart() {
-    this.items = [];
-  }
+    if (!this.items.length) {
+      alert("Your shopping cart is already empty");
+      return;
+    }
 
-  applyDiscount(amount) {
-    return parseFloat((this.discountPercentage / 100 * amount).toFixed(2))
+    // this might be the first time the confirm method is introduced. I don't remember if we are using it in the music player project
+    let isCartCleared = confirm(
+      "Are you sure you want to clear all items from your shopping cart?"
+    );
+
+    if (isCartCleared) {
+      productsContainer.innerHTML = "";
+      totalNumberOfItems.textContent = 0;
+      cartSubTotal.textContent = 0;
+      cartTaxes.textContent = 0;
+      cartTotal.textContent = 0;
+      this.items = [];
+    }
   }
 
   calculateTaxes(amount) {
-    return parseFloat((this.taxRate / 100 * amount).toFixed(2))
+    return parseFloat(((this.taxRate / 100) * amount).toFixed(2));
   }
 
   calculateTotal() {
```

Comment:
If we want to use the suggestion above about adding `this.total` to the constructor:

```js
  calculateTotal() {
    const subTotal = this.items.reduce((total, item) => total + item.price, 0);
    const tax = this.calculateTaxes(subTotal);
    const total = subTotal + tax;
    this.total = total;
    cartSubTotal.textContent = `$${subTotal.toFixed(2)}`;
    cartTaxes.textContent = `$${tax}`;
    cartTotal.textContent = `$${total.toFixed(2)}`;
    return this.total;
  }
```

Reply:
added that change 👍